sandbox/apparmor/apparmor.go

```diff
@@ -56,6 +57,221 @@ func ValidateNoAppArmorRegexp(s string) error {
+// Eg. ["/snap/core/**", "/snap/*/core/**"] where '/snap/' would be the subpattern common
+// to both excludePattern, and '*' would be the first character after the common subpattern
+// in the second excludePattern.
+// This is because there are no apparmor rules that can fulfill both requirements.
+// For /snap/[^c]** -> It will also match /snap/a/core/** which should be excluded by
+// the second pattern.
+// For /snap/[^c*]** -> It will also exclude access to /snap/a/a that should be allowed
+// as it is not explicitly excluded by any pattern
+//
+// When the '*' is used to exclude suffixes, like in ["/*.bin"], rules should be generated
+// in a reverse way:
+//
+// /*[^n]{,/**} rw,
+// /*[^i]n{,/**} rw,
+// /*[^b]in{,/**} rw,
+// /*[^.]bin{,/**} rw,
+//
+// While generating those rules is technically possible, it will make the logic way more
+// complex, thus the function would just return an error if a pattern of this kind is found.
+// This functionality can be added in a subsequent PR if needed in the future
+func GenerateAAREExclusionPatterns(excludePatterns []string, opts *AAREExclusionPatternsOptions) (string, error) {
+	if len(excludePatterns) == 0 {
+		return "", fmt.Errorf("no patterns provided")
+	}
+	seen := map[string]bool{}
+	for _, patt := range excludePatterns {
+		// check for duplicates
+		if seen[patt] {
+			return "", errors.New("exclude patterns contain duplicates")
+		}
+		seen[patt] = true
+
+		// check if it is at least legnth 2
+		if len(patt) < 2 {
+			return "", errors.New("exclude patterns must be at least length 2")
+		}
+
+		// check that it starts as an absolute path
+		if patt[0] != '/' {
+			return "", errors.New("exclude patterns must be absolute filepaths")
+		}
+
+		// TODO: should we also validate that the only character in the pattern
+		// from AARE is "*" ?
+	}
+	if opts == nil {
+		opts = &AAREExclusionPatternsOptions{}
+	}
+	return generateAAREExclusionPatternsGenericImpl(excludePatterns, opts)
+
+}
+
+func generateAAREExclusionPatternsGenericImpl(excludePatterns []string, opts *AAREExclusionPatternsOptions) (string, error) {
+
+	// Find the length of longest pattern (size)
+	size := 0
+	for _, pattern := range excludePatterns {
+		if len(pattern) > size {
+			size = len(pattern)
+		}
+	}
+
+	// Find the longest prefix common to ALL patterns.
+	commonPrefix := ""
+findCommonPrefix:
+	for charInd := 0; charInd < size; charInd++ {
+		for _, pattern := range excludePatterns {
+			if (charInd >= len(pattern)) || (pattern[charInd] != excludePatterns[0][charInd]) {
+				break findCommonPrefix
+			}
+		}
+		commonPrefix = excludePatterns[0][:charInd+1]
+	}
+
+	// This loop will iterate over the length of the longest excludePattern
+	// (charInd = 1..size), generating an apparmor rule on each iteration
+	// for the corresponding subpatterns, understanding as such, the first
+	// (charInd+1) characters of the excludePatterns.
+	builder := &strings.Builder{}
+	for charInd := 1; charInd < size; charInd++ {
```

Comment:
since we have the common prefix now, should this start from the rune right after the common prefix then?

Reply:
Even if we have the common prefix, we still need to write the apparmor rules for the common prefix, what is done by 

```
		} else if len(subpatterns) == 1 {
			res := fmt.Sprintf("%s%s[^%s]**%s\n", opts.Prefix, subpatternPrefix, subpatterns[subpatternPrefix], opts.Suffix)
			builder.WriteString(res)
```

As in the previous comments, we could move the apparmor rules generation for the common prefix out of the main loop (`for charInd := 1; charInd < size; charInd++`), however it will require to duplicate some checks (suffixes, '*' and '*/').